design/baremetal-operator/bmc-events.md

```diff
@@ -0,0 +1,146 @@
+kind: BMCEventSubscription
+metadata:
+  name: worker-1-events
+spec:
+   hostRef: ostest-worker-1
+   targetURI: https://events.apps.corp.example.com/webhook
+   headerRef: webhookBridgeAuth
+   context: “SomeUserContext”
+status:
+  errorMessage: ""
+  errorCount: 0
+  subscriptionID: aa618a32-9335-42bc-a04b-20ddeed13ade
+```
+
+- A BMCEventSubscription resource represents a subscription to the events generated
+  by a specific BMC.
+- Ironic will manage configuring the subscription using a new API for managing them.
+- The BMCEventSubscription will maintain a reference to a BareMetalHost.
+- The BMCEventSubscription will reside in the same namespace as the referenced
+  BareMetalHost.
+- The BMCEventSubscription will maintain a reference to the Ironic
+  subscription ID.
+- The BMCEventSubscription will allow injection of headers using a
+  headerRef to a secret, for example to provide basic auth
+  credentials.
+- The baremetal-operator binary will be expanded to include 2
+  reconcilers with dedicated controller/reconcile loops for
+  BareMetalHost and BMCEventSubscriptions.
+
+### Open Questions
+
+### Risks and Mitigations
+
+#### Thundering herd
+
+Large numbers of events across large numbers of BareMetalHosts could generate a
+lot of traffic. Users can control how many events their webhook receives by
+configuring the alert thresholds out of band.
+
+### Dependencies
+
+[Ironic Eventing API](https://storyboard.openstack.org/#!/story/2008366)
+needs to be complete.
+
+### Test Plan
+
+There is some existing POC code for working with Redfish Events; we
+could build on this to implement a test framework for BMC events. We could
+also consider modifying sushy-tools to support emulated eventing.
+
+### Upgrade / Downgrade Strategy
+
+Not required as this is a new API being introduced
+
+### Alternatives
+
+#### Configurable events and thresholds
+
+This API is for subscribing to events of a pre-defined type. In cases
+where no particular type is available, users would need to configure it
+out-of-band. For example, one may want to have a TemperatureOver40C
+alert that monitors the enclosure's temperature.
+
+The Redfish standard itself does not seem to have a way to specify
+specific alerts and thresholds. For example, to receive an alert when
+the temperature exceeds 40C, one would need to configure this manually
+according to the vendor's reccomendations.
+
+Vendors, however, do provide vendor-specific ways to configure these
+thresholds, but it's hard to abstract to a neutral interface. For
+example, here is a [Dell example for temperature](https://www.dell.com/support/manuals/en-jm/idrac9-lifecycle-controller-v4.x-series/idrac9_4.00.00.00_redfishapiguide_pub/temperature?guid=guid-5a798111-407b-485d-b6fb-7d6e367d4ad4&lang=en-us).
+
+In the short term, Ironic has no plans to abstract the various vendor
+implementations (if they exist at all).
+
+## References
+
+- [Ironic Eventing API](https://storyboard.openstack.org/#!/story/2008366)
```

Comment:
We also need to update this to https://storyboard.openstack.org/#!/story/2009061

Reply:
Done